From 50751075eaa745f82526b15832af1957ea1a7379 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Enrique=20Oca=C3=B1a=20Gonz=C3=A1lez?= <eocanha@igalia.com>
Date: Tue, 4 Dec 2018 11:59:17 +0000
Subject: [PATCH] omxvideodec: fix deadlock on downstream EOS

Wake the _drain condition when downstream signals GST_FLOW_EOS to
prevent the upstream streaming thread to keep waiting forever.

This scenario can be triggered when seeking near EOS.
---
 omx/gstomxvideodec.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 6034d53..c8af722 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -1598,6 +1598,10 @@ flow_error:
           gst_event_new_eos ());
       gst_pad_pause_task (GST_VIDEO_DECODER_SRC_PAD (self));
       self->started = FALSE;
+      if (self->draining) {
+        self->draining = FALSE;
+        g_cond_broadcast (&self->drain_cond);
+      }
     } else if (flow_ret < GST_FLOW_EOS) {
       GST_ELEMENT_ERROR (self, STREAM, FAILED,
           ("Internal data stream error."), ("stream stopped, reason %s",
-- 
2.17.0

